### ISSUE TITLE
使用Albert-CRF准确率很低

### ISSUE THREAD

**Issue by YestLife** [bojone/bert4keras#70]
keras2.2.0，TensorFlow1.10，bert4keras0.4.2
些许疑问
读了源码，在**bert.py**中**transformer_block**函数，大概是在200+行
逻辑是model==“albert“  --->  block_sharing==True ---> layers = layers or [] 
复用layers = layers的时候， 
attention_name = 'Encoder-%d-MultiHeadSelfAttention' % (i + 1)
feed_forward_name = 'Encoder-%d-FeedForward' % (i + 1)
永远只会存在Encoder-1-MultiHeadSelfAttention，而不会有encoder2以上

**Comment by YestLife** [comment]
> keras2.2.4，TensorFlow1.10，bert4keras0.4.2
> 
> 如题，读了你的源码之后，觉得问题是在**bert.py**中**transformer_block**函数，大概是在200+行
> 逻辑是model==“albert“ ---> block_sharing==True ---> layers = layers or []
> 但是，复用layers = layers的时候，
> attention_name = 'Encoder-%d-MultiHeadSelfAttention' % (i + 1)
> feed_forward_name = 'Encoder-%d-FeedForward' % (i + 1)
> 永远只会存在Encoder-1-MultiHeadSelfAttention，而不会有encoder2以上
> 所以你在**variable_mapping**中使用了**try：except**来避免这个错误
> try:
> self.model.get_layer('Encoder-%d-MultiHeadSelfAttention' % (i + 1))
> except ValueError:
> continue
> 
> 可，只有一层transformer并不能有效提升下游模型准确率
> 这该如何修改
> 我无法避免遇到这个问题

**Comment by bojone** [comment]
不知道你说什么。

albert本来就只有一个block，本来就是同一个block反复使用的，你用albert“只有一层transformer并不能有效提升下游模型准确率”，那关我什么事...

你要是觉得不够那你别用albert，用普通的bert就是了。

**Comment by YestLife** [comment]
感谢苏神，Albert的block反复用已经明白，之前没理解，我重新组织一下语言描述一下我的问题。

中文NER任务，keras框架，数据量3W+，version：tf-1.10，keras-2.2.0，bert4keras-0.4.6

**当均使用BERT时**，在keras-bert下一次epoch能达到80%+，在bert4keras下加载却是30%+，所以有点困惑，是在哪个步骤出了问题。**是否是layers层中层的问题**

我认为，keras-bert和bert4keras仅仅只是加载模型上有区别，下游的模型是一样的。
贴一下代码

`       # 下面一行是bert4keras的写法
        BERT = build_bert_model(config_path=config_path,
                                  checkpoint_path=checkpoint_path,
                                  model='bert',
                                  return_keras_model=True)
        # 下面一行是keras-bert的写法
        BERT = load_trained_model_from_checkpoint(config_path,checkpoint_path,seq_len, trainable=False)
        # bert4keras中使用trainable
        for _lay in BERT.layers:
            _lay.trainable = False
        # CRF层，import keras_contrib
        z = Bidirectional(GRU(bi_gru_size, return_sequences=True))(BERT.output)
        crf = CRF(label_len, sparse_target=False, learn_mode='join')
        pred = crf(z)
        model = Model(inputs=BERT.input, outputs=pred)
        model.compile(loss=crf_loss, optimizer='adam', metrics=[crf.accuracy])
        model_checkpoint = ModelCheckpoint(
            filepath=model_save_path,
            monitor='val_loss',
            save_best_only=True,
            save_weights_only=True)
        # 整合callbacks
        callbacks = [model_checkpoint]
        self.model.summary()
        hist = self.model.fit_generator(self.train_generator.forfit(),
                                steps_per_epoch=len(self.train_generator),
                                epochs=10,
                                callbacks=callbacks,
                                validation_data=self.dev_generator.forfit(),
                                validation_steps=len(self.dev_generator),
                                )
        # 存模型
        self.model.save_weights(model_save_path)`

**Comment by bojone** [comment]
因为你用到了CRF，这个问题说起来比较复杂。

bert4keras跟keras-bert不同的一点是：两者的sequence mask机制不一样。keras-bert继承了keras原有的mask机制，而你用的crf也是keras官方写的，所以两者可以适配。

但我觉得keras的mask机制并不好用，所以bert4keras没有沿用keras的mask机制，你直接套用keras的crf，它算准确率的时候事实上把padding部分也算上了，而这部分的准确率是全错的，所以拉低了整体的准确率。

解决方法是可以使用笔者自己实现的crf，参考：https://kexue.fm/archives/5542

**Comment by YestLife** [comment]
正在尝试中，我有一个疑问
如果用keras官方搭建CRF层，
但是model.compile的时候用自己写的accuracy和loss，callbacks也添加accuracy进去
可行吗

**Comment by YestLife** [comment]
> 正在尝试中，我有一个疑问
> 如果用keras官方搭建CRF层，
> 但是model.compile的时候用自己写的accuracy和loss，callbacks也添加accuracy进去
> 可行吗

这种做法应该是有效的，但是自写的callback【accuracy】的y_pred总是少一个维度，
例y_pred（：，128），y_true（：，128，13）

**Comment by YestLife** [comment]
基本已解决
需要重写一个Accuracy（callback）
目前的问题是，epoch=1是准确的，但epoch=2会突变
层数不能过高
谢谢苏神提供思路
先关闭问题了

**Comment by YestLife** [comment]
详情请见作者的范例
实测有效
https://github.com/bojone/bert4keras/blob/master/examples/task_sequence_labeling_ner_crf.py
